pyteal/ast/opup.py

```diff
@@ -0,0 +1,163 @@
+    Note: the available budget must be high enough to execute the budget
+    increase code prior to invoking ensure_budget() or maximize_budget().
+
+    Example:
+        .. code-block:: python
+
+            opup = OpUp(OpUpMode.OnCall)
+            program_with_opup = Seq(
+                ...,
+                opup.ensure_budget(Int(1000)),
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
```

Comment:
@algoidurovic and reviewers:  How essential is it to consider apps invoking `ensure_budget` > 1 per invocation?

I raise the use case because I don't know if it's been considered.  And from what I can imagine, the use case _may_ bring substantive PR changes.

* Concerns with multiple invocations:
  * It _may_ be worthwhile to define op up functionality as `@Subroutine`s to minimize program size.  In the PR's current form, the `self` reference prevents `@Subroutine` usage.  Though it's _possible_ to imagine a `@Subroutine` compliant revision. 
  * I can _imagine_ it'd be preferable for `OpUpMode.OnCall` to return the created app ID.  Retaining the app ID reduces opcode cost + transaction fees for subsequent op up calls.  I can imagine a few ways to support the use case though they'll all involve changing the API.
* Provided the use case I'm raising isn't deemed _essential_, I prefer to resolve without changes and retain a simpler approach for now.

Reply:
Ah, I realize I missed `self.target_app_id_slot.store(Gitxn[0].created_application_id()),`.  

It seems an attempt is made to cache the created app ID.  However, unless I missed it, I don't see an attempt to read the cached ID.

I'm open to a caching approach.  Though I feel the mode name (or at least docs) should reflect the caching behavior because the semantics are non-obvious and may not match the caller's intent.

Reply:
Summarizing a verbal with @algoidurovic:
* Igor to consider:  Reworking to support defining as `@Subroutine`.  If that doesn't seem like a good fit, Igor suggested we can point out to the app developer how to wrap invocations in docs.
* We agreed returning the created app ID is out of scope.  It'd be less error prone for the app developer to create the app ID out-of-band and use `Explicit` mode.

Reply:
To your first point, my position is that wrapping is a better option because the overhead is low (in terms of both implementation time/complexity and efficiency) but doesn't prevent the user from using the standard approach, which has less slot usage and fewer arguments. I'll add to the docs as soon as the code is no longer in flux. Open to feedback here.

Reply:
@algoidurovic Got it - I haven't considered the topic further.  Let's see how the docs look.  Leaving the thread open as a visual reminder.

Reply:
As of https://github.com/algorand/pyteal/pull/274/commits/07bd6d25338c3f6da89b38dd9bfcebaceb777e34, the docs briefly mention wrapping the logic to define as `@Subroutine`. 

I'm _mostly_ comfortable as is.  If program size becomes an issue due to repeated opup calls per app invocation, then we can revisit.  Resolving.